src/Storages/HDFS/StorageHDFSCluster.cpp

```diff
@@ -140,6 +134,29 @@ QueryProcessingStage::Enum StorageHDFSCluster::getQueryProcessingStage(
 }
 
 
+void StorageHDFSCluster::createIteratorAndCallback(ContextPtr context) const
+{
+    cluster = context->getCluster(cluster_name)->getClusterWithReplicasAsShards(context->getSettingsRef());
+
+    iterator = std::make_shared<HDFSSource::DisclosedGlobIterator>(context, uri);
```

Comment:
do you need to have iterator as member? you just copy it into the callback that's all the usage I see.